File: implement-shell-tools/wc/wc.mjs
import { program } from "commander";
import { promises as fs } from "node:fs";

program.name("wc");
program.description("simple wc clone");
program.option("-w, --words", "print word count");
program.option("-l, --lines", "print line count");
program.option("-c, --bytes", "print byte count");
program.argument("[paths...]", "file(s) to process");
program.parse();

const options = program.opts();
const paths = program.args;

const noFlags = !options.words && !options.lines && !options.bytes;
const showWords = options.words || noFlags;
const showLines = options.lines || noFlags;
const showBytes = options.bytes || noFlags;

function count(content) {
  const lines = content.split("\n").length - 1;
  const words = content.trim().split(/\s+/).filter(Boolean).length;
  const bytes = Buffer.byteLength(content);

  return { lines, words, bytes };
}

function formatCounts({ lines, words, bytes }, label = "") {
  const output = [];

  if (showLines) {
    output.push(lines.toString().padStart(8));
  }

  if (showWords) {
    output.push(words.toString().padStart(8));
  }

  if (showBytes) {
    output.push(bytes.toString().padStart(8));
  }

  if (label) {
    output.push(label);
  }

  return output.join(" ");
}

async function processFile(path) {
  try {
    const stat = await fs.stat(path);

    if (stat.isDirectory()) {
      console.error(`${path}: Is a directory`);
      return;
    }

    const content = await fs.readFile(path, { encoding: "utf-8" });
    const counts = count(content);
    const output = formatCounts(counts, path);

    console.log(output);
  } catch (err) {
    console.error(`${path}: ${err.message}`);
  }
}

try {
  for (const path of paths) {
    await processFile(path);
  }
} catch (err) {
  console.error(err.message);
}

File: implement-shell-tools/wc/wc.py
import argparse

parser = argparse.ArgumentParser(prog="wc",
                                 description="simple wc clone")

parser.add_argument("-w",
                    "--words",
                    dest="show_words",
                    action="store_true",
                    help="print word count")

parser.add_argument("-l",
                    "--lines",
                    dest="show_lines",
                    action="store_true",
                    help="print line count")

parser.add_argument("-c",
                    "--bytes",
                    dest="show_bytes",
                    action="store_true",
                    help="print byte count")

parser.add_argument("path")

args = parser.parse_args()

if not args.show_lines and not args.show_words and not args.show_bytes:
    args.show_lines = True
    args.show_words = True
    args.show_bytes = True

with open(args.path, "rb") as file:
    content = file.read()

word_count = str(len(content.split()))
line_count = str(content.count(b"\n"))
byte_count = str(len(content))

outputs = []

if args.show_lines:
    outputs.append(line_count)

if args.show_words:
    outputs.append(word_count)

if args.show_bytes:
    outputs.append(byte_count)

outputs.append(args.path)

output = "\t".join(outputs)
print(output)
